### PR TITLE
Gql Base Class

### PR DESCRIPTION
**Pull request by ajeet-smpx**
Split `Gql` into `Gql` (Base), `GqlApi` (Derived) & `GqlSchema` (Derived).

for https://github.com/smartprix/vutils/issues/10

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -378,38 +378,35 @@ declare module 'gqutils' {
...
+		enums?: gqlEnumMap;
+	}
+
 	interface ExecOptions {
 		context?: any;
 		cache?: {key: string, ttl?: number, forceUpdate?: boolean};
 		variables?: {[key: string]: any};
-		schemaName?: string;
-		requestOptions?: {headers?: {[key: string]: string}, cookies?: {[key: string]: string}};
+		requestOptions?: Pick<RequestOptions, 'headers' | 'cookies'>;
```
Comment (rohit-gohri) [inline index.d.ts]: allow any other key too (`...other`)

File: index.d.ts [hunk 2]
```diff
@@ -457,4 +447,23 @@ declare module 'gqutils' {
 		arg: (arg: any, opts?: string[] | {pick?: string[]}) => string;
 		toGqlArg: typeof toGqlArg;
 	}
+
+	class GqlApi<FragmentsMap = gqlFragmentMap, EnumsMap = gqlEnumMap> extends Gql<gqlFragmentMap, gqlEnumMap> {
```
Comment (rohit-gohri) [inline index.d.ts]: Types passed should not be default but the input
```suggestion
	class GqlApi<FragmentsMap = gqlFragmentMap, EnumsMap = gqlEnumMap> extends Gql<FragmentsMap, EnumsMap> {
```

File: index.d.ts [hunk 3]
```diff
@@ -457,4 +447,23 @@ declare module 'gqutils' {
...
+		constructor(api: _cacheOpts & ApiInput);
+
+		/** **NOTE**: Override this method to use your own http client */
+		static postRequest(url: string, opts: {body: any} & Omit<RequestOptions, 'endpoint'>): Promise<response>;
+	}
+
+	class GqlSchema<FragmentsMap = gqlFragmentMap, EnumsMap = gqlEnumMap> extends Gql<gqlFragmentMap, gqlEnumMap> {
```
Comment (rohit-gohri) [inline index.d.ts]: Types passed should not be default but the input
```suggestion
	class GqlSchema<FragmentsMap = gqlFragmentMap, EnumsMap = gqlEnumMap> extends Gql<FragmentsMap, EnumsMap> {
```

File: index.d.ts [hunk 4]
```diff
@@ -457,4 +447,23 @@ declare module 'gqutils' {
 		arg: (arg: any, opts?: string[] | {pick?: string[]}) => string;
 		toGqlArg: typeof toGqlArg;
 	}
+
+	class GqlApi<FragmentsMap = gqlFragmentMap, EnumsMap = gqlEnumMap> extends Gql<gqlFragmentMap, gqlEnumMap> {
+		constructor(api: _cacheOpts & ApiInput);
+
+		/** **NOTE**: Override this method to use your own http client */
+		static postRequest(url: string, opts: {body: any} & Omit<RequestOptions, 'endpoint'>): Promise<response>;
```
Comment (rohit-gohri) [inline index.d.ts]: Allow other?

Comment (rohit-gohri) [inline index.d.ts]: return value is actually data and not response anymore

File: src/Gql.js [hunk 5]
```diff
@@ -1,107 +1,27 @@
...
-
-	return execute(
-		schema,
-		document,
-		rootValue,
-		context,
-		variables,
-	);
-}
+class GqlApiError extends Error {}
```
Comment (rohit-gohri) [inline src/Gql.js]: Error classes can be moved to GqlApi and GqlSchema itself?

File: src/index.js [hunk 6]
```diff
@@ -1,6 +1,65 @@
+import {isEmpty} from 'lodash';
+import getFieldNames from 'graphql-list-fields';
+import Gql, {GqlApiError} from './Gql';
+import GqlApi from './GqlApi';
+import GqlSchema from './GqlSchema';
+
+async function postRequest(url, {headers, cookies, body, token} = {}) {
```
Comment (rohit-gohri) [inline src/index.js]: move to separate (`defaultPostRequest`) file and import from there.

File: src/index.js [hunk 7]
```diff
@@ -1,6 +1,65 @@
...
+	// eslint-disable-next-line global-require
+	const response = require('sm-utils').Connect
```
Comment (rohit-gohri) [inline src/index.js]: Remove global require, no longer needed

**Review by rohit-gohri — CHANGES_REQUESTED** [review]
Add section in Readme on how to use just the GqlApi class without importing from index and use in client side.

File: src/Gql.js [hunk 8]
```diff
@@ -253,9 +84,7 @@ class Gql {
 			query = `query { ${query} }`;
 		}
 
-		const result = this._api ?
-			await this._execApi(query, {variables, requestOptions}) :
-			await this._execGraphql(query, {context, variables});
+		const result = await this._getQueryResult(query, {context, variables, requestOptions});
```
Comment (rohit-gohri) [inline src/Gql.js:84]: add this to types too? as an abstract function

**Review by rohit-gohri — APPROVED** [review]

**Comment by hit12369** [comment]
Is this meant to be used from the frontend? It isn't mentioned what to import. I guess package.json would need to be changed too.

**Comment by rohit-gohri** [comment]
> Is this meant to be used from the frontend? It isn't mentioned what to import. I guess package.json would need to be changed too.

It is meant for clients. Can be used from frontend too though. A `broswer` field in package.json?

There is an example in Readme

**Comment by hit12369** [comment]
> > Is this meant to be used from the frontend? It isn't mentioned what to import. I guess package.json would need to be changed too.
> 
> It is meant for clients. Can be used from frontend too though. A `broswer` field in package.json?
> 
> There is an example in Readme

Yes, a browser field. Readme doesn't specify how to import the class. Also a sample implementation with `fetch` would be nice (don't import by default, also mention in readme).

**Comment by rohit-gohri** [comment]
> Yes, a browser field. Readme doesn't specify how to import the class. Also a sample implementation with `fetch` would be nice (don't import by default, also mention in readme).


Done

**Pull request merged**